### ISSUE TITLE
~/aur vs manual aura -Ayu

### ISSUE THREAD

**Issue by jshuping** [pigmonkey/spark#4]
Hi again, wondering if you have any advice on what I'm doing wrong, or what secret you may have for this. Here is what I have done:
- haven't run ansible in about 3 weeks
- since then, I _have_ run `sudo aura -Ayu` which has updated several pkgs
- I noticed you had made some playbook changes, pulled those, ran ansible
- and now a handful of aur packages got downgraded, since my ~/aur hadn't changed since 3 weeks ago:

```
    $  grep downgraded /var/log/pacman.log 
    [2015-10-27 09:02] [ALPM] downgraded chromium-pepper-flash (1:19.0.0.226-1 -> 1:19.0.0.207-1)
    [2015-10-27 09:02] [ALPM] downgraded firejail (0.9.32-1 -> 0.9.30-1)
    [2015-10-27 09:02] [ALPM] downgraded virtualbox-ext-oracle (5.0.8-1 -> 5.0.6-1)
    [2015-12-08 10:39] [ALPM] downgraded aura-bin (1.3.4-1 -> 1.3.2.1-1)
```

Am I incorrectly expecting the playbook to be fairly idempotent like how I use puppet, and run it multiple times? In other words are you only running ansible once on a new system?
Or are you updating your ~/aur somehow?  Guess I could just nuke ~/aur, but I think I found another idempotency issue so I thought I'd ask this question. Cheers!

**Comment by pigmonkey** [comment]
Interesting. I haven't experienced that. So far I've only rerun targeted tasks via tags when needed. I've probably just happened to avoid rerunning any of the stuff involving AUR packages. I definitely want all of this to be idempotent.

It seems like the `--needed` we pass to makepkg ought to avoid downgrading, but I guess not.

So the easy solution would be to register the result of the cower tasks, and then only run the makepkg tasks if the previous cower task was succesfull. If cower was skipped, makepkg would be skipped too. That would avoid this problem (and speed up reruns a bit). But it also means that if the makepkg task were to fail the first time you run ansible, they would never be attempted again (unless you deleted the relevant package directory from `~/aur`).

**Comment by jshuping** [comment]
That could work.. what if we only run cower+makepkg if the package isn't already installed? Then we wouldn't face the issue you called out, and reruns would still speed up a bit.

Or maybe, what if instead of cower, we git clone'd the various PKGBUILD repos, and could git pull them if they already existed.. optionally cloning with my ssh key so I can edit/submit stuff later? That might be nice..

**Comment by pigmonkey** [comment]
Checking if the package is installed would look like this:

```
- name: Check if pmount is installed
  command: "pacman -Qq pmount"
  failed_when: false
  register: installed
  tags:
    - aur
    - packages

- name: Download pmount
  command: cower -dq pmount
           chdir=/home/{{ user.name }}/{{ aur.dir }}
           creates=/home/{{ user.name }}/{{ aur.dir }}/pmount
  when: installed.rc != 0
  tags:
    - aur
    - packages
  become: yes
  become_user: "{{ user.name }}"

- name: Build and install pmount
  command: "{{ aur.makepkg }} chdir=/home/{{ user.name }}/{{ aur.dir }}/pmount"
  when: installed.rc != 0
  tags:
    - aur
    - packages
  become: yes
  become_user: "{{ user.name }}"
```

That seems ridiciously long to install one package.

What's needed is an Ansible module to install AUR packages (that doesn't depend on yet another AUR helper).

**Comment by jshuping** [comment]
I agree that is wayyy long, and personally I think even the current amount of lines required to install a single AUR package is pretty long too.  

What about collapsing it all int a single command?

```
if ! pacman -Qq pmount; then [ -d ~/aur/pmount ] || cower -dq pmount ; makepkg -s ; fi
```

Or, what about we write our own little shell script which ansible could install/use for this set of steps?

Any obvious downsides to either of those approaches? I can't think of any yet, but this coffee is pretty weak this morning.

**Comment by pigmonkey** [comment]
I haven't built an Ansible module before but [it seems easy enough](https://github.com/pigmonkey/ansible-aur). Plugging that into spark will let us cleanup all the cruft.

**Comment by pigmonkey** [comment]
Fixed with 2d0c544683483ffaac45359979e05ef54c22e212.
